test/crowdsale/PostDeliveryRefundableCrowdsale.test.js

```diff
@@ -0,0 +1,103 @@
+const time = require('../helpers/time');
+const shouldFail = require('../helpers/shouldFail');
+const { ether } = require('../helpers/ether');
+
+const BigNumber = web3.BigNumber;
+
+require('chai')
+  .use(require('chai-bignumber')(BigNumber))
+  .should();
+
+const PostDeliveryRefundableCrowdsaleImpl = artifacts.require('PostDeliveryRefundableCrowdsaleImpl');
+const SimpleToken = artifacts.require('SimpleToken');
+
+contract('PostDeliveryRefundableCrowdsale', function ([_, investor, wallet, purchaser]) {
+  const rate = new BigNumber(1);
+  const tokenSupply = new BigNumber('1e22');
+  const goal = ether(100);
+
+  before(async function () {
+    // Advance to the next block to correctly read time in the solidity "now" function interpreted by ganache
+    await time.advanceBlock();
+  });
+
+  beforeEach(async function () {
+    this.openingTime = (await time.latest()) + time.duration.weeks(1);
+    this.closingTime = this.openingTime + time.duration.weeks(1);
+    this.afterClosingTime = this.closingTime + time.duration.seconds(1);
+    this.token = await SimpleToken.new();
+    this.crowdsale = await PostDeliveryRefundableCrowdsaleImpl.new(
+      this.openingTime, this.closingTime, rate, wallet, this.token.address, goal
+    );
+    await this.token.transfer(this.crowdsale.address, tokenSupply);
+  });
+
+  context('after opening time', function () {
+    beforeEach(async function () {
+      await time.increaseTo(this.openingTime);
+    });
+
+    context('with bought tokens below the goal', function () {
+      const value = goal.sub(1);
+
+      beforeEach(async function () {
+        await this.crowdsale.buyTokens(investor, { value: value, from: purchaser });
+      });
+
+      it('does not immediately deliver tokens to beneficiaries', async function () {
+        (await this.crowdsale.balanceOf(investor)).should.be.bignumber.equal(value);
+        (await this.token.balanceOf(investor)).should.be.bignumber.equal(0);
+      });
+
+      it('does not allow beneficiaries to withdraw tokens before crowdsale ends', async function () {
+        await shouldFail.reverting(this.crowdsale.withdrawTokens(investor));
+      });
+
+      context('after closing time and finalization', function () {
+        beforeEach(async function () {
+          await time.increaseTo(this.afterClosingTime);
+          await this.crowdsale.finalize();
```

Comment:
Doesn't testing both conditions at the same time result in less than 100% coverage?

Reply:
Could you expand a bit on what you mean by 'two conditions'?

Reply:
I mean having
- `context('after closing time and finalization'`

Instead of 
- `context('after closing time'`
  - `context('after finalization'`

Reply:
Line coverage will remain the same because those conditions are already accounted for in some other test (probably `TimedCrowdsale` and `FinalizableCrowdsale`). We are building on top of those tests by only testing for what it is this crowdsale modifiers, as opposed to going all out and testing the complete interaction space.

Reply:
But this directly affects the branch coverage of the one function that was added in this PR. I don't see how it can be 100%.

Reply:
Ah, I see what you mean. There are 2 `require` statements, but we don't need 4 cases to cover them all: `solidity-coverage` only checks that each individual `require` is hit. I did add 4 scenarios in the test, but only to showcase that withdrawals being rejected before finalization is independent of the goal being reached or not.

Re. closing time, the reason why that works is that `RefundablePostBlahBlah` doesn't care about time, only finalization (`Refundable` is a `FinalizableCrowdsale`).

Reply:
Ah, I understand now. Thanks.